datafusion/storage/src/read.rs

```diff
@@ -0,0 +1,149 @@
+            Included(i) => *i,
+            Excluded(i) => i + 1,
+            Unbounded => 0,
+        };
+        match value.end_bound() {
+            Included(i) => Self::Bounded(first..(i + 1)),
+            Excluded(i) => Self::Bounded(first..*i),
+            Unbounded => Self::Offset(first),
+        }
+    }
+}
+
+/// Options for a read request, such as range
+#[derive(Debug, Default, Clone)]
+pub struct StorageReadOptions {
+    /// Request will succeed if the `StorageFileMetadata::e_tag` matches
+    /// otherwise returning [`Error::Precondition`]
+    ///
+    /// See <https://datatracker.ietf.org/doc/html/rfc9110#name-if-match>
+    ///
+    /// Examples:
+    ///
+    /// ```text
+    /// If-Match: "xyzzy"
+    /// If-Match: "xyzzy", "r2d2xxxx", "c3piozzzz"
+    /// If-Match: *
+    /// ```
+    pub if_match: Option<String>,
+    /// Request will succeed if the `StorageFileMetadata::e_tag` does not match
+    /// otherwise returning [`Error::NotModified`]
+    ///
+    /// See <https://datatracker.ietf.org/doc/html/rfc9110#section-13.1.2>
+    ///
+    /// Examples:
+    ///
+    /// ```text
+    /// If-None-Match: "xyzzy"
+    /// If-None-Match: "xyzzy", "r2d2xxxx", "c3piozzzz"
+    /// If-None-Match: *
+    /// ```
+    pub if_none_match: Option<String>,
+    /// Request will succeed if the object has been modified since
+    ///
+    /// <https://datatracker.ietf.org/doc/html/rfc9110#section-13.1.3>
+    pub if_modified_since: Option<DateTime<Utc>>,
+    /// Request will succeed if the object has not been modified since
+    /// otherwise returning [`Error::Precondition`]
+    ///
+    /// Some stores, such as S3, will only return `NotModified` for exact
+    /// timestamp matches, instead of for any timestamp greater than or equal.
+    ///
+    /// <https://datatracker.ietf.org/doc/html/rfc9110#section-13.1.4>
+    pub if_unmodified_since: Option<DateTime<Utc>>,
+    /// Request transfer of only the specified range of bytes
+    /// otherwise returning [`Error::NotModified`]
+    ///
+    /// <https://datatracker.ietf.org/doc/html/rfc9110#name-range>
+    pub range: Option<ReadRange>,
+    /// Request a particular object version
+    pub version: Option<String>,
+}
+
+#[async_trait]
+pub trait StorageFileRead: Send + Sync + 'static {
+    async fn read(&mut self, ctx: &StorageContext) -> Result<Bytes>;
+}
+
+pub struct StorageFileReader {
+    meta: StorageFileMetadata,
+    inner: Box<dyn StorageFileRead>,
+}
+
+/// Expose public API for reading from a file
+impl StorageFileReader {
+    pub fn metadata(&self) -> &StorageFileMetadata {
+        &self.meta
+    }
+
+    pub fn into_futures_reader(self) -> StorageFileFuturesReader {
+        todo!()
+    }
+
+    pub fn into_stream(self) -> StorageFileBytesStream {
+        todo!()
+    }
+
+    pub fn into_tokio_reader(self) -> StorageFileTokioReader {
+        todo!()
+    }
+}
+
+/// Adapter to allow using `futures::io::AsyncRead` with `StorageFileReader`
```

Comment:
I have the same question about read as write -- to all these APIs require copying the data (aka take the data as `&[u8]`)?

One of the neat things we do with the object store API and parquet reader when reading strings in `Utf8View`  is that the same `Bytes` buffer that is returned from `object_store` is converted into a parquet Buffer and then eventually forms the underlying Arrow Buffer of the returned array

IN other words there is no copying data around once it comes back from object store

Reply:
Most thing is similar to https://github.com/apache/datafusion/pull/15018#discussion_r1986341562, we will have:

- zero cost `StorageFileRead`
- `StorageFileReader::into_stream`
- `StorageFileReader::into_futures_read` and `StorageFileReader::into_tokio_read`